src/main/java/io/debezium/connector/spanner/db/metadata/SchemaRegistry.java

```
@@ -89,38 +96,68 @@ private boolean validate(TableId tableId, List<Column> rowType) {
+                    column.isNullable(), dialect);
 
-        SpannerSchema newSchema;
-        if (this.changeStream.isWatchedAllTables()) {
-            newSchema = schemaDao.getSchema(timestamp);
-        }
-        else {
-            newSchema = schemaDao.getSchema(timestamp, this.changeStream.getTables());
         }
+        SpannerSchema newSchema = builder.build();
         if (this.spannerSchema == null) {
             this.spannerSchema = newSchema;
             return;
         }
         this.spannerSchema = SchemaMerger.merge(this.spannerSchema, newSchema);
+        LOGGER.info("Schema registry has been updated to stale timestamp {}", timestamp);
+
+    }
+
+    @VisibleForTesting
+    void forceUpdateSchema(@Nullable TableId tableId, Timestamp updatedTimestamp, @Nullable List<Column> rowTypes) {
+        try {
+            this.timestamp = updatedTimestamp;
+            this.changeStream = schemaDao.getStream(timestamp, streamName);
+
+            if (this.changeStream == null) {
+                throw new IllegalStateException("Change stream doesn't exist at timestamp: "
+                        + streamName + ", " + timestamp);
+            }
+
+            SpannerSchema newSchema;
+            if (this.changeStream.isWatchedAllTables()) {
+                newSchema = schemaDao.getSchema(timestamp);
+            }
+            else {
+                newSchema = schemaDao.getSchema(timestamp, this.changeStream.getTables());
+            }
+            if (this.spannerSchema == null) {
+                this.spannerSchema = newSchema;
+                return;
+            }
+            this.spannerSchema = SchemaMerger.merge(this.spannerSchema, newSchema);
+        }
+        catch (SpannerException e) {
+            if (e.getMessage().contains("has exceeded the maximum timestamp staleness") && e.getErrorCode().equals(ErrorCode.FAILED_PRECONDITION)) {
+                updateSchemaFromStaleTimestamp(tableId, timestamp, rowTypes);
+            }
```

Do we want to throw this error if the error doesn't match?

---

Done.